Client crash with ender chest

```
[B#402]  [15:31:39] [main/FATAL] [VF]: Minecraft ran into a problem! Report saved to: /home/bluexin/.technic/modpacks/the-faerun-heroes/crash-reports/crash-2019-01-22_15.31.39-client.txt
[B#402]  ---- Minecraft Crash Report ----
[B#402]  // Uh... Did I do that?
[B#402]  
[B#402]  Time: 2019-01-22 15:31:39 CET
[B#402]  Description: Error executing task
[B#402]  
[B#402]  java.util.concurrent.ExecutionException: java.lang.IndexOutOfBoundsException: Index: 63, Size: 63
[B#402]      at java.util.concurrent.FutureTask.report(FutureTask.java:122)
[B#402]      at java.util.concurrent.FutureTask.get(FutureTask.java:192)
[B#402]      at net.minecraft.util.Util.runTask(SourceFile:531)
[B#402]      at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1087)
[B#402]      at net.minecraft.client.Minecraft.run(Minecraft.java:3940)
[B#402]      at net.minecraft.client.main.Main.main(SourceFile:123)
[B#402]      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[B#402]      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[B#402]      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[B#402]      at java.lang.reflect.Method.invoke(Method.java:498)
[B#402]      at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
[B#402]      at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
[B#402]  Caused by: java.lang.IndexOutOfBoundsException: Index: 63, Size: 63
[B#402]      at java.util.ArrayList.rangeCheck(ArrayList.java:657)
[B#402]      at java.util.ArrayList.get(ArrayList.java:433)
[B#402]      at net.minecraft.inventory.Container.getSlot(Container.java:120)
[B#402]      at net.minecraft.inventory.Container.setAll(Container.java:527)
[B#402]      at net.minecraft.client.network.NetHandlerPlayClient.handleWindowItems(NetHandlerPlayClient.java:1191)
[B#402]      at net.minecraft.network.play.server.SPacketWindowItems.processPacket(SourceFile:50)
[B#402]      at net.minecraft.network.play.server.SPacketWindowItems.processPacket(SourceFile:12)
[B#402]      at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:20)
[B#402]      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[B#402]      at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[B#402]      at net.minecraft.util.Util.runTask(SourceFile:529)
[B#402]      ... 9 more
```

--- a/build.gradle
+++ b/build.gradle
@@ -165,5 +165,6 @@ compileKotlin {
     kotlinOptions {
         javaParameters = true
         jvmTarget = 1.8
+        freeCompilerArgs += "-Xjvm-default=enable"
     }
 }

--- a/coremod/src/main/java/be/bluexin/rpg/core/BlueRPGHooks.java
+++ b/coremod/src/main/java/be/bluexin/rpg/core/BlueRPGHooks.java
@@ -20,12 +20,14 @@ package be.bluexin.rpg.core;
 import be.bluexin.rpg.events.CreatePlayerContainerEvent;
 import be.bluexin.rpg.events.CreatePlayerInventoryEvent;
 import be.bluexin.rpg.events.LivingEquipmentPostChangeEvent;
+import be.bluexin.rpg.events.OpenEnderChestEvent;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.ContainerPlayer;
 import net.minecraft.inventory.EntityEquipmentSlot;
+import net.minecraft.inventory.IInventory;
 import net.minecraftforge.common.MinecraftForge;
 
 @SuppressWarnings("unused")
@@ -48,5 +50,11 @@ public class BlueRPGHooks {
 
     public static void equipmentPostChangeHook(EntityLivingBase entity, EntityEquipmentSlot slot) {
         MinecraftForge.EVENT_BUS.post(new LivingEquipmentPostChangeEvent(entity, slot));
+    }
+
+    public static void openEnderChestGui(EntityPlayer player, IInventory chest) {
+        OpenEnderChestEvent evt = new OpenEnderChestEvent(player, chest);
+        MinecraftForge.EVENT_BUS.post(evt);
+        if (!evt.isCanceled()) player.displayGUIChest(chest);
     }
 }

--- a/coremod/src/main/java/be/bluexin/rpg/core/BlueRPGTransformer.java
+++ b/coremod/src/main/java/be/bluexin/rpg/core/BlueRPGTransformer.java
@@ -49,6 +49,7 @@ public class BlueRPGTransformer implements IClassTransformer, Opcodes {
     static {
         transformers.put("net.minecraft.entity.player.EntityPlayer", BlueRPGTransformer::transformPlayer);
         transformers.put("net.minecraft.entity.EntityLivingBase", BlueRPGTransformer::transformELB);
+        transformers.put("net.minecraft.block.BlockEnderChest", BlueRPGTransformer::transformBEC);
     }
 
     private static byte[] transformPlayer(byte[] basicClass) {
@@ -139,6 +140,26 @@ public class BlueRPGTransformer implements IClassTransformer, Opcodes {
                         }
                         next = next.getNext();
                     }
+
+                    return true;
+                }
+        ));
+    }
+
+    private static byte[] transformBEC(byte[] basicClass) {
+        // func_180639_a(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/entity/player/EntityPlayer;Lnet/minecraft/util/EnumHand;Lnet/minecraft/util/EnumFacing;FFF)Z
+        final MethodSignature onBlockActivated = new MethodSignature("onBlockActivated", "func_180639_a", "(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/state/IBlockState;Lnet/minecraft/entity/player/EntityPlayer;Lnet/minecraft/util/EnumHand;Lnet/minecraft/util/EnumFacing;FFF)Z");
+        final MethodSignature displayGUIChest = new MethodSignature("displayGUIChest", "func_71007_a", "(Lnet/minecraft/inventory/IInventory;)V");
+
+        return transform(basicClass, onBlockActivated, "BlockEnderChest openGUI", combine(
+                (node) -> node.getOpcode() == INVOKEVIRTUAL && displayGUIChest.matches((MethodInsnNode) node),
+                (method, node) -> {
+                    InsnList l = new InsnList();
+                    l.add(new MethodInsnNode(INVOKESTATIC, ASM_HOOKS, "openEnderChestGui",
+                            "(Lnet/minecraft/entity/player/EntityPlayer;Lnet/minecraft/inventory/IInventory;)V", false));
+
+                    method.instructions.insertBefore(node, l);
+                    method.instructions.remove(node);
 
                     return true;
                 }

--- a/coremod/src/main/java/be/bluexin/rpg/events/OpenEnderChestEvent.java
+++ b/coremod/src/main/java/be/bluexin/rpg/events/OpenEnderChestEvent.java
@@ -1,0 +1,37 @@
+/*
+ * Copyright (C) 2019.  Arnaud 'Bluexin' Solé
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package be.bluexin.rpg.events;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.IInventory;
+import net.minecraftforge.event.entity.player.PlayerEvent;
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+
+@Cancelable
+public class OpenEnderChestEvent extends PlayerEvent {
+    private final IInventory chest;
+
+    public OpenEnderChestEvent(EntityPlayer player, IInventory chest) {
+        super(player);
+        this.chest = chest;
+    }
+
+    public IInventory getChest() {
+        return chest;
+    }
+}

--- a/src/main/kotlin/be/bluexin/rpg/EventHandler.kt
+++ b/src/main/kotlin/be/bluexin/rpg/EventHandler.kt
@@ -24,6 +24,7 @@ import be.bluexin.rpg.containers.RPGEnderChestContainer
 import be.bluexin.rpg.events.CreatePlayerContainerEvent
 import be.bluexin.rpg.events.CreatePlayerInventoryEvent
 import be.bluexin.rpg.events.LivingEquipmentPostChangeEvent
+import be.bluexin.rpg.events.OpenEnderChestEvent
 import be.bluexin.rpg.gear.IRPGGear
 import be.bluexin.rpg.gear.WeaponAttribute
 import be.bluexin.rpg.gear.WeaponType
@@ -51,10 +52,8 @@ import net.minecraft.client.renderer.tileentity.TileEntityItemStackRenderer
 import net.minecraft.entity.EntityLivingBase
 import net.minecraft.entity.item.EntityPainting
 import net.minecraft.entity.player.EntityPlayer
-import net.minecraft.inventory.ContainerChest
 import net.minecraft.inventory.ContainerPlayer
 import net.minecraft.inventory.EntityEquipmentSlot
-import net.minecraft.inventory.InventoryEnderChest
 import net.minecraft.item.ItemStack
 import net.minecraft.nbt.NBTTagByte
 import net.minecraft.util.ResourceLocation
@@ -69,7 +68,6 @@ import net.minecraftforge.event.entity.EntityEvent
 import net.minecraftforge.event.entity.EntityJoinWorldEvent
 import net.minecraftforge.event.entity.living.*
 import net.minecraftforge.event.entity.player.CriticalHitEvent
-import net.minecraftforge.event.entity.player.PlayerContainerEvent
 import net.minecraftforge.event.entity.player.PlayerInteractEvent
 import net.minecraftforge.event.entity.player.PlayerPickupXpEvent
 import net.minecraftforge.event.world.BlockEvent
@@ -379,15 +377,9 @@ object CommonEventHandler {
 
     @SubscribeEvent
     @JvmStatic
-    fun openContainer(event: PlayerContainerEvent.Open) {
-        val c = event.container
-        when (c) {
-            is ContainerChest -> {
-                if (c.lowerChestInventory is InventoryEnderChest) {
-                    GuiHandler.open(RPGEnderChestContainer.NAME, event.entityPlayer, BlockPos.ORIGIN)
-                }
-            }
-        }
+    fun openEnderchest(event: OpenEnderChestEvent) {
+        event.isCanceled = true
+        GuiHandler.open(RPGEnderChestContainer.NAME, event.entityPlayer, BlockPos.ORIGIN)
     }
 
     @ConfigProperty("security", "Whether to protect paintings from breaking")
@@ -448,7 +440,6 @@ object ClientEventHandler {
         when (g) {
             is GuiInventory -> event.gui =
                     GuiRpgInventory((g.inventorySlots as ContainerImpl).container as RPGContainer)
-//            is GuiChest -> event.gui = GuiRpgEnderInventory(RPGEnderChestContainer(Minecraft().player))
         }
     }
 

--- a/src/main/kotlin/be/bluexin/rpg/Proxy.kt
+++ b/src/main/kotlin/be/bluexin/rpg/Proxy.kt
@@ -23,6 +23,7 @@ import be.bluexin.rpg.blocks.BlockCaster
 import be.bluexin.rpg.blocks.BlockEditor
 import be.bluexin.rpg.blocks.BlockGatheringNode
 import be.bluexin.rpg.containers.ContainerEditor
+import be.bluexin.rpg.containers.RPGEnderChestContainer
 import be.bluexin.rpg.entities.*
 import be.bluexin.rpg.gear.*
 import be.bluexin.rpg.items.DebugExpItem
@@ -133,6 +134,7 @@ open class CommonProxy : CoroutineScope {
         EggItem
 
         EntitySkillProjectile.Companion // This is needed for it's custom serializer
+        RPGEnderChestContainer.Companion // Classloading FTW
     }
 
     private fun registerEntities() {

--- a/src/main/kotlin/be/bluexin/rpg/gui/GuiRpgInventory.kt
+++ b/src/main/kotlin/be/bluexin/rpg/gui/GuiRpgInventory.kt
@@ -156,7 +156,7 @@ class PlayerLayout(player: RPGContainer.InventoryWrapperPlayer) {
     val bags = ComponentVoid(180, 92).apply {
         player.bags.forEachIndexed { index, slot ->
             add(ComponentSlot(slot, index * 18, 0).apply {
-                background.add(ComponentSprite(Textures.BAG_BG, 0, -1))
+                background.add(ComponentSprite(Textures.BAG_BG, -1, -1))
             })
         }
     }

